### PR TITLE
29901: GraphQL mutation updateCartItems return data is incorrect when removing multiple items

### LINKED ISSUE #29901: GraphQL mutation uptdateCartItems return data is incorrect when removing multiple items
### Preconditions (*)
1. Magento version: 2.4.0

### Steps to reproduce (*)

1. Add a set of items to cart (minimum 2)
2. Using the graphQL mutation `updateCartItems` set multiple or all of the item quantities to 0

query
```
mutation {
  updateCartItems(
    input: {
      cart_id: {{cartID}}",
      cart_items: [
        {
          cart_item_id: 1
          quantity: 0
        },
        {
          cart_item_id: 2
          quantity: 0
        },
      ]
    }
  ) {
    cart {
      items {
        id
        product {
          name
        }
        quantity
      }
      prices {
        grand_total{
          value
          currency
        }
      }
    }
  }
}
```
### Expected result (*)
1. Expected cart data where all items that were set to a quantity of 0 to be removed

### Actual result (*)

1. The cart data that is returned will remove 1 item. The rest of the items will still be in the cart.
2. Even though All items were set to 0. The cart returned with all items minus 1.
3. The Items are still removed from cart, but an extra query to get the cart was required to get the correct cart data.
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [X] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by sudheers-kensium**
GraphQL mutation uptdateCartItems return data is incorrect when removing multiple items

### Description (*)

1. Add a set of items to cart (minimum 2)
2. Using the graphQL mutation updateCartItems set multiple or all of the item quantities to 0

query

``mutation {
  updateCartItems(
    input: {
      cart_id: {{cartID}}",
      cart_items: [
        {
          cart_item_id: 1
          quantity: 0
        },
        {
          cart_item_id: 2
          quantity: 0
        },
      ]
    }
  ) {
    cart {
      items {
        id
        product {
          name
        }
        quantity
      }
      prices {
        grand_total{
          value
          currency
        }
      }
    }
  }
}
``
3. Expected cart data where all items that were set to a quantity of 0 to be removed

### Fixed Issues (if relevant)

1. Fixes magento/magento2#29901

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by sudheers-kensium** [comment]
@magento run all tests

**Comment by sudheers-kensium** [comment]
@magento run Functional Tests CE

**Comment by sudheers-kensium** [comment]
@magento run Functional Tests EE

**Comment by sudheers-kensium** [comment]
@magento run Functional Tests B2B

**Comment by sudheers-kensium** [comment]
@magento run Functional Tests EE

File: app/code/Magento/QuoteGraphQl/Model/Resolver/UpdateCartItems.php [hunk 1]
```diff
@@ -77,6 +77,7 @@ public function resolve(Field $field, $context, ResolveInfo $info, array $value
         try {
             $this->updateCartItems->processCartItems($cart, $cartItems);
             $this->cartRepository->save($cart);
+            $cart = $this->getCartForUser->execute($maskedCartId, $context->getUserId(), $storeId);
```
Comment (rogyar) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/UpdateCartItems.php]: Could we try to achieve the same result without an additional database query? I mean, we already have all the details about the shopping cart. Most likely, we don't need an additional DB fetch, but the recalculation process that happens upon the DB fetch.

File: dev/tests/integration/testsuite/Magento/Checkout/_files/quote_with_multiple_items.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+)->addProduct(
+    $product,
+    2
+);
+$quote->save();
+$quote->addProduct(
+    $productTwo,
+    2
+);
+$quote->save();
```
Comment (rogyar) [inline dev/tests/integration/testsuite/Magento/Checkout/_files/quote_with_multiple_items.php]: Please, avoid using deprecated methods in the code.

File: dev/tests/integration/testsuite/Magento/Checkout/_files/quote_with_multiple_items.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    $accountManagement->getPasswordHash('password')
+)->setReservedOrderId(
+    'test_order_1'
+)->setCustomerEmail(
+    'aaa@aaa.com'
+)->addProduct(
+    $product,
+    2
+);
+$quote->save();
```
Comment (rogyar) [inline dev/tests/integration/testsuite/Magento/Checkout/_files/quote_with_multiple_items.php]: Please, avoid using deprecated methods in the code.

File: dev/tests/integration/testsuite/Magento/Checkout/_files/quote_with_multiple_items_saved.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+);
+$quoteRepository->save($quote);
+
+/** @var \Magento\Quote\Model\QuoteIdMask $quoteIdMask */
+$quoteIdMask = \Magento\TestFramework\Helper\Bootstrap::getObjectManager()
+    ->create(\Magento\Quote\Model\QuoteIdMaskFactory::class)
+    ->create();
+$quoteIdMask->setQuoteId($quote->getId());
+$quoteIdMask->setDataChanges(true);
+$quoteIdMask->save();
```
Comment (rogyar) [inline dev/tests/integration/testsuite/Magento/Checkout/_files/quote_with_multiple_items_saved.php]: Please, avoid using deprecated methods in the code.

File: dev/tests/integration/testsuite/Magento/Checkout/_files/quote_with_multiple_items_saved_rollback.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+use Magento\TestFramework\Workaround\Override\Fixture\Resolver;
+
+/** @var $objectManager \Magento\TestFramework\ObjectManager */
+$objectManager = \Magento\TestFramework\Helper\Bootstrap::getObjectManager();
+$quote = $objectManager->create(\Magento\Quote\Model\Quote::class);
+$quote->load('test_order_1', 'reserved_order_id')->delete();
+
+/** @var \Magento\Quote\Model\QuoteIdMask $quoteIdMask */
+$quoteIdMask = $objectManager->create(\Magento\Quote\Model\QuoteIdMask::class);
+$quoteIdMask->delete($quote->getId());
```
Comment (rogyar) [inline dev/tests/integration/testsuite/Magento/Checkout/_files/quote_with_multiple_items_saved_rollback.php]: Please, avoid using deprecated methods in the code.

File: dev/tests/integration/testsuite/Magento/Checkout/_files/quote_with_multiple_items.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
+<?php
```
Comment (rogyar) [inline dev/tests/integration/testsuite/Magento/Checkout/_files/quote_with_multiple_items.php]: If we create new fixtures for GraphQL api-functional tests, I would strongly recommend using a dedicated place (directory) for that. Please, check the following fixtures as the example 

https://github.com/magento/magento2/tree/a5a98dfa56cac8cdb60e15568c10578f522be3f3/dev/tests/integration/testsuite/Magento/GraphQl/Quote/_files

(please, take a look at the files path)

**Review by rogyar — CHANGES_REQUESTED** [review]
Hi @sudheers-kensium. Thanks for your contribution. That a solid piece of work. Please, check my suggestions below. 
It's strongly recommended to keep best practices even when we are working with the fixtures. In the future, we will need to support them as a part of the codebase. 

Thank you!

**Comment by sudheers-kensium** [comment]
@magento run all tests

**Comment by sudheers-kensium** [comment]
@magento run all tests

**Comment by sudheers-kensium** [comment]
@magento run all tests

**Comment by sudheers-kensium** [comment]
@magento run Functional Tests B2B

**Comment by sudheers-kensium** [comment]
@magento run Functional Tests B2B

**Comment by sudheers-kensium** [comment]
@magento run Functional Tests B2B

**Comment by sudheers-kensium** [comment]
@magento run Functional Tests B2B

**Review by rogyar — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-8423](https://jira.corp.magento.com/browse/ENGCOM-8423) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
Hi @rogyar & @sudheers-kensium 
Thanks for your contribution and collaboration!
I have tried to reproduce this issue and seems to be not reproducible. Kindly cross-check below steps I followed please!

1. Created a cart using GraphQL request.
2. Added 3 products with 2 quantity each as shown in the below screenshot.
Products added to Cart
3. Updated the cart with zero quantity for 2 of the products among 3, and the screenshot is below
update to cart
4. GraphQL queried that retrieves the cart products and list.
cart after updating

As defined in the [issue description](https://github.com/magento/magento2/issues/30220#issue-710370985), didn't notice any issue as such. Zeroed quantity is successfully ignored in the cart queried response.
Tested it in the 2.4-develop Magento instance, and I request you to kindly check if the issue is still existing. Since the issue was noticed many months back, it might have fixed with other PRs.

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Alfa** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Alfa** [comment]
This issue is in the discussion, will be processed soon.
Thanks for your patience @rogyar & @sudheers-kensium

**Comment by engcom-Alfa** [comment]
Hi @sudheers-kensium & @rogyar 
Thanks for your collaboration and contribution here.

Based on the [above-comment ](https://github.com/magento/magento2/pull/30648#issuecomment-971522834)and [today's comment update in issue](https://github.com/magento/magento2/issues/29901#issuecomment-1225561582), we are currently closing this PR ticket.
In case, if you find the same issue in the Magento code `2.4-develop` version, please re-update the _steps to reproduce_ here, or if there is any configurations or anything required to done to reproduce it. Or you may report a new PR as well. Based on that, we will re-consider it.

Thanks

**Pull request closed without merging**